pkg/buildqueue/buildqueue.go

```diff
@@ -0,0 +1,568 @@
+
+	q.mutex.Lock()
+	b.waiting++
+	metric.Inc()
+	for !b.condition(q) {
+		q.waitingBuilders.Wait()
+	}
+
+	metric.Dec()
+	b.waiting--
+	b.running++
+	q.mutex.Unlock()
+}
+
+// finish signals that a build of this buildGroup has completed
+func (b *buildGroup) finish(q *BuildQueue) {
+	q.mutex.Lock()
+	b.running--
+	q.mutex.Unlock()
+
+	q.waitingBuilders.Broadcast()
+}
+
+// BuildQueue is a queueing system for builds
+type BuildQueue struct {
+	// name is the name of the build queue
+	name string
+
+	// mutex protects access to the buildStatus map and for the
+	// waitingBuilders sync.Condition in waitForBuildCondition()
+	mutex lock.Mutex
+
+	// buildStatus contains the status of all builders separated by UUID
+	buildStatus buildStatusMap
+
+	// workerBuildQueue is a buffered channel that contains all scheduled
+	// builds
+	workerBuildQueue chan Builder
+
+	// stopWorker is used to stop worker threads when the build queue is
+	// shutdown. Workers will run until this channel is closed.
+	stopWorker chan struct{}
+
+	// regularBuilds accounts for running regular builds
+	regularBuilds *buildGroup
+
+	// exclusiveBuilds accounts for running exclusive builds
+	exclusiveBuilds *buildGroup
+
+	// waitingBuilders is used to wake up builders waiting for their
+	// condition to be met
+	waitingBuilders *sync.Cond
+}
+
+// NewBuildQueue returns a new build queue
+func NewBuildQueue(name string) *BuildQueue {
+	q := &BuildQueue{
+		name:             name,
+		buildStatus:      buildStatusMap{},
+		workerBuildQueue: make(chan Builder, buildQueueSize),
+		stopWorker:       make(chan struct{}, 0),
+		regularBuilds: &buildGroup{
+			condition: func(q *BuildQueue) bool {
+				return q.exclusiveBuilds.waiting == 0 && q.exclusiveBuilds.running == 0
+			},
+		},
+		exclusiveBuilds: &buildGroup{
+			condition: func(q *BuildQueue) bool {
+				return q.exclusiveBuilds.running == 0 && q.regularBuilds.running == 0
+			},
+		},
+	}
+
+	q.waitingBuilders = sync.NewCond(&q.mutex)
+
+	nWorkers := numWorkerThreads()
+	for w := 0; w < nWorkers; w++ {
+		go q.runWorker()
+	}
+
+	return q
+}
+
+// Stop stops the build queue and terminates all workers
+func (q *BuildQueue) Stop() {
+	close(q.stopWorker)
+
+	waitForCleanup := sync.WaitGroup{}
+
+	q.mutex.Lock()
+	for _, status := range q.buildStatus {
+		waitForCleanup.Add(1)
+		go func(b Builder) {
+			q.Drain(b)
+			waitForCleanup.Done()
+		}(status.builder)
+	}
+	q.mutex.Unlock()
+
+	waitForCleanup.Wait()
+}
+
+// numWorkerThreads returns the number of worker threads to use
+func numWorkerThreads() int {
+	ncpu := runtime.NumCPU()
+
+	if ncpu < minWorkerThreads {
+		return minWorkerThreads
+	}
+	return ncpu
+}
+
+// Buildable is an object that is buildable
+type Buildable interface {
+	// GetUUID must return a unique UUID of the object
+	GetUUID() string
+}
+
+// Builder is an object that can build itself. A builder must also be Buildable
+type Builder interface {
+	Buildable
+
+	// Build must build object
+	Build() error
```

Comment:
Nitpick: reorder the methods to match the order in which they are called:
- `BuildQueued`
- `BuildsDequeued`
- `Build`

I would also document that order in the type's doc above.
Also document that both `BuildQueued` and `BuildsDequeued` are always called once a builder is enqueued. But `Build` may not be called, e.g. if it's removed or drained.